#Data Integration Process

Downloading required Text Preprocessing Libraries

In [ ]:
!pip install texthero

In [ ]:
!pip3 install spacy==3.4

Importing the libraries

In [ ]:
import pandas as pd
import texthero as hero

/usr/local/lib/python3.8/dist-packages/torch/cuda/__init__.py:497: UserWarning: Can't initialize NVML
  warnings.warn("Can't initialize NVML")
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


#Data Integration Process












Imported combined and clean preprocessing data which i have uploaded earlier on github and performing validation check if the dataset is imbalance.

In [ ]:
df = pd.read_csv('combined_dataset.csv')
df.head()

text  category
0  as a woman you shouldn t complain about cleani...         0
1  boy dats cold tyga dwn bad for cuffin dat hoe ...         1
2  dawg you ever fuck a bitch and she start to cr...         1
3                             she look like a tranny         1
4  the shit you hear about me might be true or it...         1

Changing the category column name to label.... 🥰 just following a common layout

In [ ]:
df.rename(columns = {'category':'label'}, inplace = True)

Viewing overall observations grouped by labels to check whether the data is imbalanced or not

In [ ]:
df['label'].value_counts()

0    30584
1    18680
2    13485
Name: label, dtype: int64

Ohh!! 😲 Its seems like data is imbalanced.

As Our lead Vishu Kalier and co-leader Caterina Bonan has suggested that we need maximum 30000 approx records in each category but if you observe in category 1 and 2, records are not close to 30000. so to balance this data we need to add more records to 1 and 2 category only
 

##Here is the list of the datasets which gonna use:


1.   Clean mendeley hatespeech dataset partA
2.   Toxic Comment Classification Challenge
3.   Malignant comments clean






###Clean mendeley hatespeech dataset partA

Downloading Dataset from Github

In [ ]:
!wget https://raw.githubusercontent.com/OmdenaAI/aswan-egypt-detecting-hate-nlp/main/src/data/english/clean-datasets/mendeley_dataset/clean_mendeley_hatespeech_dataset_partA.csv?token=GHSAT0AAAAAAB37HHF6VZVEQFCZSMRH6KHOY7LPKLA

--2023-02-14 12:11:24--  https://raw.githubusercontent.com/OmdenaAI/aswan-egypt-detecting-hate-nlp/main/src/data/english/clean-datasets/mendeley_dataset/clean_mendeley_hatespeech_dataset_partA.csv?token=GHSAT0AAAAAAB37HHF6VZVEQFCZSMRH6KHOY7LPKLA
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 93211830 (89M) [text/plain]
Saving to: ‘clean_mendeley_hatespeech_dataset_partA.csv?token=GHSAT0AAAAAAB37HHF6VZVEQFCZSMRH6KHOY7LPKLA’

clean_mendeley_hate 100%[===================>]  88.89M   188MB/s    in 0.5s    

2023-02-14 12:11:30 (188 MB/s) - ‘clean_mendeley_hatespeech_dataset_partA.csv?token=GHSAT0AAAAAAB37HHF6VZVEQFCZSMRH6KHOY7LPKLA’ saved [93211830/93211830]



Importing the dataset to pandas dataframe

In [ ]:
df1 = pd.read_csv('clean_mendeley_hatespeech_dataset_partA.csv')
df1.head()

Unnamed: 0       corpus_name  \
0           0  mendeley_dataset   
1           1  mendeley_dataset   
2           2  mendeley_dataset   
3           3  mendeley_dataset   
4           4  mendeley_dataset   

                                        raw_sentence  label  \
0  `- This is not ``creative``.  Those are the di...      0   
1  `  :: the term ``standard model`` is itself le...      0   
2    True or false, the situation as of March 200...      0   
3   Next, maybe you could work on being less cond...      0   
4               This page will need disambiguation.       0   

                             clean_sentence_training  \
0  This creative Those dictionary definition term...   
1  term standard model le NPOV I think wed prefer...   
2  True false situation March  such A Saudi propo...   
3  Next maybe could work le condescending suggest...   
4                      This page need disambiguation   

                                  clean_sentence_EDA  
0  This creative Those dictionary definition term...  
1  term standard model le NPOV I think wed prefer...  
2  True false situation March  such A Saudi propo...  
3  Next maybe could work le condescending suggest...  
4                      This page need disambiguation

Selecting required columns as we are going to preprocess this data later

In [ ]:
df1 = df1[['raw_sentence','label']]

Grouping the total observation by label 0 denote non risky and 2 denote risky:

In [ ]:
df1['label'].value_counts()

0    81297
1    12183
Name: label, dtype: int64

As you see we have enough risky category data i.e label 1 data to fix imbalance problem in risky category

so we will filter out only label 1 data 

Renaming column name to make sure we follow common layout

In [ ]:
df1.rename(columns = {'raw_sentence':'text'}, inplace = True)

Filtering out label 1 data which we will appened to imbalanced label 2 category

In [ ]:
df1= df1.loc[df1.label > 0]

Renaming label name to 2 to make sure it doesnt get append tp label 1 category i.e potenially risky

In [ ]:
df1['label'] = df1['label'] = 2

Overview of clean mendeley hatespeech_dataset partA

In [ ]:
df1.head()

text  label
52   `  :Click on my ``Annoying Users`` link! I gue...      2
60           Iraq is not good  ===  ===  USA is bad         2
121  `  Buddha - ``Some suggest that victims should...      2
207    ____ fuck off you little asshole. If you wan...      2
351       i have a dick, its bigger than yours! hahaha      2

##Toxic Comment Classification Challenge Dataset

In [ ]:
df2 = pd.read_csv('Toxic Comment Classification Challenge.csv', index_col = [0])
df2.head()

corpus_name  \
0  Toxic Comment Classification Challenge   
1  Toxic Comment Classification Challenge   
2  Toxic Comment Classification Challenge   
3  Toxic Comment Classification Challenge   
4  Toxic Comment Classification Challenge   

                                        raw_sentence  label  \
0  Explanation\nWhy the edits made under my usern...      0   
1  D'aww! He matches this background colour I'm s...      0   
2  Hey man, I'm really not trying to edit war. It...      0   
3  "\nMore\nI can't make any real suggestions on ...      0   
4  You, sir, are my hero. Any chance you remember...      0   

                             clean_sentence_training  \
0  Explanation Why edits made username Hardcore M...   
1  D aww He match background colour I seemingly s...   
2  Hey man I really trying edit war It guy consta...   
3  More I make real suggestion improvement I wond...   
4              You sir hero Any chance remember page   

                                  clean_sentence_EDA  
0  Explanation Why edits made username Hardcore M...  
1  D aww He match background colour I seemingly s...  
2  Hey man I really trying edit war It guy consta...  
3  More I make real suggestion improvement I wond...  
4              You sir hero Any chance remember page

In [ ]:
df2 = df2[['raw_sentence','label']]

In [ ]:
df2.rename(columns = {'raw_sentence':'text'}, inplace = True)

Grouping the total observation by label 0 denote non risky, 1 denote potenially risky and 2 denote risky:

In [ ]:
df2['label'].value_counts()

0    58815
2     6208
1      512
Name: label, dtype: int64

As we see here we are interested only in label 2 and 1. 
We will perform similar steps as above where we will filter out records and add it to imbalanced dataset

In [ ]:
df2= df2.loc[df2.label > 0]

Overview of filtered Toxic Comment Classification Challenge Dataset

In [ ]:
df2.head()

text  label
6        COCKSUCKER BEFORE YOU PISS AROUND ON MY WORK      2
12  Hey... what is it..\n@ | talk .\nWhat is it......      2
16  Bye! \n\nDon't look, come or think of comming ...      2
42  You are gay or antisemmitian? \n\nArchangel WH...      2
43           FUCK YOUR FILTHY MOTHER IN THE ASS, DRY!      2

Performing the validation check if the data is filtered out correctly or not

In [ ]:
df2['label'].value_counts()

2    6208
1     512
Name: label, dtype: int64

Appending the raw sentence datasets

In [ ]:
df3 = df1.append(df2)

this is output of combined raw dataset

In [ ]:
df3['label'].value_counts()

2    18391
1      512
Name: label, dtype: int64

well for now we have manged to fix imbalance in risky category. now we have to fix imbalance in potenially risky category for that we are going to use Malignant comments clean Dataset where i have found desired number of records for pontenially risky category

##Malignant comments clean Dataset




In [ ]:
df4 = pd.read_csv('sample_data/malignant_comments_clean.csv')
df4.head()

corpus_name  \
0  Malignant_Comment_Classification   
1  Malignant_Comment_Classification   
2  Malignant_Comment_Classification   
3  Malignant_Comment_Classification   
4  Malignant_Comment_Classification   

                                        raw_sentence  \
0  Explanation\nWhy the edits made under my usern...   
1  D'aww! He matches this background colour I'm s...   
2  Hey man, I'm really not trying to edit war. It...   
3  "\nMore\nI can't make any real suggestions on ...   
4  You, sir, are my hero. Any chance you remember...   

                             clean_sentence_training  \
0  Explanation Why edits made username Hardcore M...   
1  D'aww He match background colour I 'm seemingl...   
2  Hey man I 'm really trying edit war It 's guy ...   
3  More I ca n't make real suggestion improvement...   
4           You sir hero Any chance remember page 's   

                                  clean_sentence_EDA  label  
0  Explanation Why edits made username Hardcore M...      0  
1  D'aww He match background colour I 'm seemingl...      0  
2  Hey man I 'm really trying edit war It 's guy ...      0  
3  More I ca n't make real suggestion improvement...      0  
4           You sir hero Any chance remember page 's      0

Here I have performed same steps as above

In [ ]:
df4['label'].value_counts()

0    123932
1     33179
2      2460
Name: label, dtype: int64

In [ ]:
df4 = df4[['raw_sentence', 'label']]

In [ ]:
df4.rename(columns = {'raw_sentence':'text'}, inplace = True)

In [ ]:
df4.head()

text  label
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

Filtering out label 1 category records

In [ ]:
df4= df4.loc[df4['label'] == 1]

In [ ]:
df4.head()

text  label
8   Sorry if the word 'nonsense' was offensive to ...      1
9   alignment on this subject and which are contra...      1
25  Radial symmetry \n\nSeveral now extinct lineag...      1
26  There's no need to apologize. A Wikipedia arti...      1
27  Yes, because the mother of the child in the ca...      1

In [ ]:
df4.label.value_counts()

1    33179
Name: label, dtype: int64

As you see we have 33179 in label 1 category which is huge if we add all label 1 records to imbalance dataset the total number records in label 1 category will be close to 50000 approx so keep 30000 records mark for each category i will take random sample from filtered dataset and add it to imbalance dataset

In [ ]:
df5 = df4.sample(frac = 0.4)

In [ ]:
df5.label.value_counts()

1    13272
Name: label, dtype: int64

Appending the combined raw sentence dataset with Malignant comments raw Dataset

In [ ]:
df6 = df3.append(df5)

As you see we have required number of records to fix imbalance dataset

In [ ]:
df6.label.value_counts()

2    18391
1    13784
Name: label, dtype: int64

Overview of combined 3 raw sentences dataset

In [ ]:
df6.head()

text  label
52   `  :Click on my ``Annoying Users`` link! I gue...      2
60           Iraq is not good  ===  ===  USA is bad         2
121  `  Buddha - ``Some suggest that victims should...      2
207    ____ fuck off you little asshole. If you wan...      2
351       i have a dick, its bigger than yours! hahaha      2

#Text Processing


> Requirements

As per the requirement of our lead Vishu Kalier and colead Caterina Bonan we were expected to remove Usernames, Urls, Punctuations and White Spaces excluding stopwords and apply lemmatization on sentences

For text preprocessing I have used Regular expression (RE) and automated text preprocessing library TextHero

You take a look at the documantation of this library

https://texthero.org/docs/api-preprocessing








In [ ]:
df6['text'] = df6['text'].apply(str)# map dataframe to string

In [ ]:
import re
df6['text'] = df6['text'].apply(lambda x : re.sub(r'@[^\s]+','', x).strip()) # removing symbols from sentences

In [ ]:
df6['text'] = df6['text'].replace("[^a-zA-Z]"," ",regex=True) # removing  Punctuations

In [ ]:
df6['text'] = hero.preprocessing.remove_urls(df6['text']) # removing urls if any

In [ ]:
df6['text'] = hero.preprocessing.lowercase(df6['text']) # converting all text to lower case

Apply lemmatization on text sentences

In [ ]:
import nltk
nltk.download('omw-1.4')

[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


True

In [ ]:
import nltk # import nltk library for lemmatization
nltk.download('wordnet')
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
def lemmatize_words(text): # creating function to lemmatize each sentence from dataframe
    words = text.split() # spliting sentence into words before lemmatiztion
    words = [lemmatizer.lemmatize(word,pos='v') for word in words] # apply lemmatiztion on each word by loop
    return ' '.join(words) # joining all lemmatize words into sentences
df6['text'] = df6['text'].apply(lemmatize_words) # applying the function

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [ ]:
df6.head()

text  label
52   click on my annoy users link i guess massachus...      2
60                         iraq be not good usa be bad      2
121  buddha some suggest that victims should be ref...      2
207  fuck off you little asshole if you want to tal...      2
351         i have a dick its bigger than yours hahaha      2

Wow!! 🙂 Finally we have completed text preprocessing, now its time to append this clean combined data to out imbalanced dataset 

In [ ]:
df7 = df.append(df6)

Overview of our model ready dataset

In [ ]:
df7

text  label
0       as a woman you shouldn t complain about cleani...      0
1       boy dats cold tyga dwn bad for cuffin dat hoe ...      1
2       dawg you ever fuck a bitch and she start to cr...      1
3                                  she look like a tranny      1
4       the shit you hear about me might be true or it...      1
...                                                   ...    ...
102984  oh know and you be do so well irishlass oh wel...      1
5966    spoilers i be think of add information to this...      1
58788   two can play this game they ll have to protect...      1
106879  myopic be not a swear word it simply mean shor...      1
139057  conspiracy no only a unionists would think tha...      1

[94924 rows x 2 columns]

As you see we managed to fix this imbalance dataset and its now well balanced dataset. we can use this dataset for model training

In [ ]:
df7['label'].value_counts()

1    32464
2    31876
0    30584
Name: label, dtype: int64

Saving the dataset

In [ ]:
df7.to_csv('model_ready_dataset1.csv')